_posts/2019-03-08-code-quality-inspection-for-clojure-with-sonarqube.md

```diff
@@ -0,0 +1,194 @@
+|foo.login                     |   68.67 |   82.76 |
+|foo.util                      |    2.89 |   20.00 |
+|------------------------------|---------|---------|
+|                    ALL FILES |   67.36 |   82.93 |
+
+
+### Lein-nvd
+
+[Lein-nvd](https://github.com/rm-hull/lein-nvd) is a dependency-checker plugin whichs checks JARS in the programs classpath for known vulnerabilites against 
+the [National Vulnerability Database](https://nvd.nist.gov/). 
+
+Output of Lein-nvd gives status of the JARs with found vulnerabilities:
+
+
+| dependency             | status|
+|------------------------|-------|
+| bcpkix-jdk15on-1.58.jar | OK    | 
+| bcprov-jdk15on-1.58.jar| CVE-2017-13098, CVE-2018-1000180, CVE-2018-1000613|
+
+It should be noted that Lein-nvd does not exactly tell which dependency or its transitive dependency pulls a vulnerable JAR file. Most time JAR name matches a dependency. The full dependency tree can be inspected by running ```lein deps :tree``` command.
+
+### Ancient
+
+[Ancient-clj](https://github.com/xsc/lein-ancient/tree/master/ancient-clj) is a plugin to check your project for outdated dependencies and plugins and  suggest updates. 
+
+Sample output looks like this:
+
+```
+[ring "1.7.1"] is available but we use "1.6.3"
+```
+
+Ancient can be started with command ```lein ancient upgrade :interactive``` which gives option to one by one decide if the upgrade is made or not. The upgrade also runs tests and persists the changes only if the tests run correctly. 
+
+Ancient cannot analyze if the new version is backwards compatible or not. Many libraries uses semantic versioning major.minor.patch (1.2.3) where minor (1.**2**.3) and patch (1.1.**2**) version changes should be backwards compatible. Whether this is really the case can differ.
+
+
+## Consolidating code quality and vulnerability findings using SonarQube
+
+[SonarQube](https://www.sonarqube.org/) is a code inspection tool which supports on the box many currently used languages like Java, Python, C++ and so on. SonarQube server have a browser GUI for browsing the findings.
+
+![SonarQube GUI](/img/code-quality-inspection-for-clojure-with-sonarqube/sonar-results.png)
+
+SonarQube unfortunately doesnt have out of a box support for Clojure. Fortunately SonarQube can be extended to support more languages with plugins. 
+
+There have been many different Clojure plugins for SonarQube but most of them have not been maintained and doesnt work with newer versions of SonarQube. When I started to look for SonarQube plugin only working one was [plugin](https://github.com/fsantiag/sonar-clojure) made by Felipe Santiago. The plugin only supported Eastwood based linting.
+
+Because plugin was open sourced and the maintainer was active I made necessary changes to the plugin which added support for Kibit, Cloverage, Lein-nvd and Ancient which are now merged. The plugin have also a minimal SonarQube docker image file for testing the plugin locally. 
```

Comment:
"Because the plugin"
"made the necessary changes"
"The plugin also has a minimal"